Fix UI not updating when thumbtable image is rejected

sayoder opened the pull request:
Fixes the following bug:
Steps to reproduce:
    - Enter darkroom view
    - Scale bottom panel UI small enough that rating overlay does not display on
      thumbnail mouseover
    - Hover over an image, and press R.
    - Observe that the image does not appear to be rejected until we
      enter another view where the rating overlay is displayed.

Previously, _image_get_infos short-circuited when a thumbnail did not have an overlay. When using keyboard shortcuts, there are situations where we still need to call _image_get_infos to appropriately update the UI.

sayoder commented:
@TurboGit - [Here](https://www.youtube.com/watch?v=_RIZYrO9K7s) is a video where I demonstrate the bug.

This video shows four tests:

- Hover over non-rejected image in lighttable, press `r`. 
  - Expect: thumbnail greyed out. This works.
- Hover over rejected image in lighttable, press `r`. 
  - Expect: thumbnail no longer greyed out. This works.
- Hover over non-rejected image in darkroom thumbnail strip, press `r`. 
  - Expect: thumbnail greyed out. 
  - Actual: thumbnail remains solid until we switch back to lighttable view -- observe that it is greyed out there, and it retains that state when we switch to darktable view again.
- Hover over rejected image in darkroom thumbnail strip, press `r`. 
  - Expect: thumbnail no longer greyed out. 
  - Actual: thumbnail remains greyed out until we switch back to lighttable view, etc. etc. etc.

sayoder commented:
@TurboGit , sorry I didn't answer your question directly. [Here](https://youtu.be/phvkdeWj4WY) is a video that shows the behavior I'm talking about. I think this behavior is intended, but I'm not sure.

When the thumbtable strip is too small, the rating overlay does not display when I mouse over a thumbnail.

sayoder commented:
@TurboGit , sorry I didn't answer your question directly. [Here](https://www.youtube.com/watch?v=OiAQnRGKuU0) is a video that shows the behavior I'm talking about in that sentence. I think this behavior is intended, but I'm not sure.

When the thumbtable strip is too small, the rating overlay does not display when I mouse over a thumbnail.

AlicVB commented on src/dtgtk/thumbnail.c:
```diff
@@ -215,8 +215,6 @@ static void _image_get_infos(dt_thumbnail_t *thumb)
     return;
-  if(thumb->over == DT_THUMBNAIL_OVERLAYS_NONE)
-    return;
 
```
That means we'll *always* read the metadata, even if not really needed. That may slow up a little thumb loading...
The ideal changes here would be to have a specific path to retrieve only reject values (because afaik it's the only data we show on thumbs without overlays)

AlicVB reviewed (changes requested):
Thanks for the PR.
You've spotted the code problem, but I think we should be able to get a change with less impact on perfs (even if current proposal is not horrible)
Ping me if you need help !

sayoder commented:
Hi @AlicVB , let me know if you think the newest commit addresses your concerns. The new logic is:

- gather the bare minimum info required to check if the image is rejected.
- if reject status has changed, update the UI class.
- if overlay is `DT_THUMBNAIL_OVERLAYS_NONE`, release cache and exit.
- continue gathering other info for other overlays

Thanks in advance!

AlicVB commented on src/dtgtk/thumbnail.c:
```diff
@@ -226,10 +224,26 @@ static void _image_get_infos(dt_thumbnail_t *thumb)
     thumb->rating = img->flags & DT_IMAGE_REJECTED
       ? DT_VIEW_REJECT
       : (img->flags & DT_VIEW_RATINGS_MASK);
+  }
```
Is there any reason to close the `if(img)` condition here (and reopen it after) ?
I think it would even break the `dt_image_cache_release` just after (in case of `OVERLAYS_NONE`)

In fact I even think that current code in master is not fully right : in the case we fail to get the image cache, it actually "reset" the `thumb->rating` to 0. I would have said that in this case, we have to let everything untouched...
So for our case that would means : 
- removing the `thumb->rating = 0` line
- let all your change inside the `if(img)` condition
- add another check after that condition to exit if no overlay

What would you think of suvh a change ? @TurboGit : I would also like your opinion as I fail to trigger a case where we don't get the image cache...

Thanks !

AlicVB reviewed (commented):
First, I'm really sorry for the extra long delay. I've completely forgotten that PR...

Your proposal is far better, but I have a more "global" interrogation here

AlicVB commented:
closing as this is fixed by #18462 now merged in master
